src/main/java/racingcar/service/RacingGameService.java

```diff
@@ -0,0 +1,34 @@
+package racingcar.service;
+
+import java.util.List;
+import racingcar.domain.RoundCount;
+import racingcar.domain.car.CarName;
+import racingcar.domain.car.Cars;
+import racingcar.dto.CarStatusDto;
+import racingcar.dto.GameResultDto;
+import racingcar.dto.RoundResultDto;
+
+public class RacingGameService {
+
+    private final GameResultPublishService gameResultPublishService;
+    private final CarFactoryService carFactoryService;
+    private final RoundExecutionService roundExecutionService;
+    private final WinnerSelectionService winnerSelectionService;
+
+    public RacingGameService(GameResultPublishService gameResultPublishService,
+                             CarFactoryService carFactoryService,
+                             RoundExecutionService roundExecutionService,
+                             WinnerSelectionService winnerSelectionService) {
+        this.gameResultPublishService = gameResultPublishService;
+        this.carFactoryService = carFactoryService;
+        this.roundExecutionService = roundExecutionService;
+        this.winnerSelectionService = winnerSelectionService;
+    }
+
```

Comment:
객체를 인자로 받아와서 생성하신 이유가 궁금해요!

```suggestion

   private final GameResultPublishService gameResultPublishService = new GameResultPublishService();
    private final CarFactoryService carFactoryService = new CarFactoryService();
    private final RoundExecutionService roundExecutionService = new RoundExecutionService();
    private final WinnerSelectionService winnerSelectionService = new WinnerSelectionService();

    public RacingGameService() {
    }
```

제 생각에는 이런식으로 인자를 줄이면 더 깔끔하고 생성자를 사용할 때 어떤 인자가 필요한지 일일이 확인하지 않아도 돼서 좋다고 생각하거든요 혹시 다른 이유가 있어서 인자로 객체를 받으신걸까요? 의견 듣고싶습니다. 👍

Reply:
의존성 주입에서 필드 주입과 생성자 주입의 차이이군요! 👍🏻 

성욱님이 제시하신 것처럼 필드를 통해서 의존 객체를 주입하면 확실히 깔끔하기는 합니다! 깔끔하다는 이유로 실제로 이전에 많은 개발자 분들이 필드주입을 사용했다고 해요!

그런데 필드 주입은 의존관계가 아주 강하게 설정됩니다. 어떤 객체를 의존할 지 한번 정하면 코드를 수정하지 않는 이상 의존성을 변경할 수 없죠.
반면 생성자 주입은 의존관계가 비교적 느슨합니다. Service객체를 생성할 때 생성자를 통해 어떤 객체를 의존할 지 정할 수 있다는 점에서 확장성이 있죠.

예로 나중에 누군가 가장 짧게 움직이는 차가 우승하는 모드도 있었으면 좋겠다고 요구사항을 말했다고 해보죠 🤔
저희는 아마 새로운 서비스 클래스를 만들 것입니다! ```ShortestMoveWinnerSelectionService```와 같이요 ㅎㅎ 

이런 경우에 생성자 주입이 이점을 가집니다. RacingGameService를 생성할 때 생성자에서 의존할 객체를 바꿔주는 것만으로 게임의 모드를 바꿀 수 있죠! 반면 필드 주입은 서비스 코드를 직접 수정하지 않으면 절대 바꿀 수 없습니다. 확장성이 떨어지는 것이죠!

사실 생성자 주입의 진짜 진가는 객체의 의존성을 관리하는 Config 클래스를 운용할 때 나타나는데요! 그런데 저는 제출 30분전까지 코드를 막 수정하다가 Config 클래스를 운용하지 못해서 생성자 주입의 의미가 떨어진 상태이네요 😭 

그러한 점에서는 성욱님 의견처럼 필드 주입을 기용했어도 괜찮았을 것 같습니다 .

어쨌든! 확장성을 고려해서 생성자 주입을 선택한 것이 제 의견이었고 성욱님의 의견도 너무 감사합니다 🙇🏻 

성욱님에게 답변을 드리면서 저도 지금 엄청 공부하게 되었네요..! 질 높은 질문 감사합니다. 제가 참고한 글 첨부 드릴게요 한번 확인해보세요! 

https://alisyabob.tistory.com/327